plugins/janus_streaming.c

```diff
@@ -3352,13 +3521,291 @@ void janus_streaming_incoming_rtp(janus_plugin_session *handle, int video, char
+										!this_session->destroyed &&
+										!this_session->stopping &&
+										this_session->started &&
+										!this_session->paused &&
+										(this_session->last_audio_rr_received > ss->last_audio_rr_sent) &&
+										(this_session->audio_rr_fraction_lost > max_fraction_lost))
+										max_fraction_lost = this_session->audio_rr_fraction_lost;
+									this_listener = g_list_next(this_listener);
+								}
+								janus_mutex_unlock(&mp->mutex);
+
+								/* update this RR message */
+								rr->rb[0].flcnpl = htonl(((uint32_t)max_fraction_lost << 24) | (ntohl(rr->rb[0].flcnpl) & 0xffffff));
+
+								/* send it */
+								ss->last_audio_rr_sent = now;
+								keep = TRUE;
+							}
+						}
+					}
+					break;
+				case RTCP_PSFB:
+				case RTCP_SR:
+				case RTCP_SDES:
+				case RTCP_BYE:
+				case RTCP_APP:
+				case RTCP_RTPFB:
+				case RTCP_XR:
+				case RTCP_FIR:
+					break;
+				default:
+					JANUS_LOG(LOG_ERR, "Unknown RTCP PT %d\n", rtcp->type);
+					/* FIXME Should we allow this to go through instead? */
+					break;
+			}
+			if(keep) {
+				/* Keep this packet */
+				if(filtered == NULL)
+					filtered = g_malloc0(total);
+				memcpy(filtered+newlen, (char *)rtcp, bytes);
+				newlen += bytes;
+			}
+			total -= bytes;
+			if(total <= 0)
+				break;
+			rtcp = (janus_rtcp_header *)((uint32_t*)rtcp + length + 1);
+		}
+
+
+		if(filtered) {
+			struct sockaddr_in dest_rtcp;
+			dest_rtcp.sin_family = AF_INET;
+			dest_rtcp.sin_port = htons(ss->audio_rtcp_port);
+			dest_rtcp.sin_addr.s_addr = ss->audio_mcast;
+			int sent = 0;
+
+			if (0 > (sent = sendto(ss->audio_rtcp_fd, buf, len, 0, &dest_rtcp, sizeof(dest_rtcp)))) {
+				JANUS_LOG(LOG_ERR, "Error sendto %d... %d (%s)\n",
+					ss->audio_rtcp_port, errno, strerror(errno));
+			} else {
+				JANUS_LOG(LOG_HUGE, "sent %d/%d bytes to %d\n",
+					sent, len, ss->audio_rtcp_port);
+			}
+		}
+	} else
+	if((video != 0) && (ss->video_rtcp_fd > -1)) {
+		JANUS_LOG(LOG_VERB, "Got Video RTCP SSRC %u\n",
+			janus_rtcp_get_sender_ssrc(buf, len));
+
+		janus_rtcp_fix_ssrc(NULL, buf, len, 1,
+			janus_rtcp_get_sender_ssrc(buf, len),
+			ss->video_ssrc);
+
+		char *filtered = NULL;
+		int newlen = 0;
+		int total = len, length = 0, bytes = 0;
+		/* Iterate on the compound packets */
+		gboolean keep = FALSE;
+		while(rtcp) {
+			keep = FALSE;
+			length = ntohs(rtcp->length);
+			if(length == 0)
+				break;
+			bytes = length*4+4;
+			switch(rtcp->type) {
+				case RTCP_PSFB:
+					/* incoming PLI or REMB */
+					if(15 == rtcp->rc) {
+						/* possibly REMB */
+						uint32_t this_remb = janus_rtcp_get_remb((char*)rtcp, length);
+						if(this_remb > 0) {
+							JANUS_LOG(LOG_HUGE, "REMB for this PeerConnection: %u\n",
+								this_remb);
+							session->video_remb = this_remb;
+
+							/* iterate through listeners and compare against last REMB
+							 * if this is a new minimum bandwidth, send it along
+							 */
+							GList *this_listener = NULL;
```

Comment:
I'm not very fond of this part either. That's a lot of code, and forces an iteration on all viewers any time we handle a REMB, which is a linear process and possibly inefficient. We worked a lot to try and remove these kind of loops in the core as well, especially in the ICE loop, so I'd rather avoid adding new ones here. There are simpler approaches that would allow us to only keep track of the "smallest" REMB value in the configured value, e.g., as monited when dealing with incoming RTP traffic instead (which is supposed to be steady most of the times).